src/views/components/trace/trace-detail-chart-list.vue

```diff
@@ -145,66 +146,111 @@ export default {
       const segmentIdGroup = [];
       const fixSpans = [];
       const segmentHeaders = [];
-      this.data.forEach((span) => {
-        if (span.parentSpanId === -1) {
-          segmentHeaders.push(span);
-        } else {
-          const index = this.data.findIndex(i => (i.segmentId === span.segmentId && i.spanId === (span.spanId - 1)));
-          if (index === -1) {
-            fixSpans.push(
-              {
-                traceId: span.traceId, segmentId: span.segmentId, spanId: span.spanId - 1, parentSpanId: span.spanId - 2, refs: [], endpointName: `VNode: ${span.segmentId}`, serviceCode: 'VirtualNode', type: 'Broken', peer: '', component: `VirtualNode: #${span.spanId - 1}`, isError: true, isBroken: true, layer: 'Broken', tags: [], logs: [],
-              },
-            );
+        this.data.forEach((span) => {
+          if (span.parentSpanId === -1) {
+            segmentHeaders.push(span);
+          } else {
+            const index = this.data.findIndex(i => (i.segmentId === span.segmentId && i.spanId === (span.spanId - 1)));
+            const fixSpanKeyContent = {
+              traceId: span.traceId,
+              segmentId: span.segmentId,
+              spanId: span.spanId - 1,
+              parentSpanId: span.spanId - 2,
+            };
+            if (index === -1 && !_.find(fixSpans, fixSpanKeyContent)) {
+              fixSpans.push(
+                {
+                  ...fixSpanKeyContent, refs: [], endpointName: `VNode: ${span.segmentId}`, serviceCode: 'VirtualNode', type: `[Broken] ${span.type}`, peer: '', component: `VirtualNode: #${span.spanId - 1}`, isError: true, isBroken: true, layer: 'Broken', tags: [], logs: [],
+                },
+              );
+            }
           }
-        }
-      });
-      segmentHeaders.forEach((span) => {
-        if (span.refs.length) {
-          span.refs.forEach((ref) => {
-            const index = this.data.findIndex(i => (ref.parentSegmentId === i.segmentId && ref.parentSpanId === i.spanId));
-            if (index === -1) {
-              for (let i = 0; i <= ref.parentSpanId; i += 1) {
-                fixSpans.push(
+        });
+        segmentHeaders.forEach((span) => {
+          if (span.refs.length) {
+            span.refs.forEach((ref) => {
+              const index = this.data.findIndex(i => (ref.parentSegmentId === i.segmentId && ref.parentSpanId === i.spanId));
+              if (index === -1) {
+                // create a known broken node.
```

Comment:
@TinyAllen The following part makes sense to me. Create the missing parent vnode, then targeting its parent to root vnode. And every time check the vnode whether exist before doing `#push`.